src/coq/constants/equtils.mli

```diff
@@ -96,3 +96,36 @@ val is_rewrite_r : types ->  bool
  * Don't consider convertible terms
  *)
 val is_rewrite : types -> bool
+
+
+(* 
+ * Information required to perform a rewrite. 
+ *)
+type rewrite_args = {
+    a : types;
+    (* x : A *)
+    x : constr;
+    (* motive P : A -> Type/Prop/Set *)
+    p : constr;
+    (* proof of P x *)
+    px : constr;
+    (* y : A *)
+    y : constr;
+    (* x = y if "<-", y = x otherwise *)
+    eq : constr;
+    (* additional arguments following equality *)
+    params : constr array;
+    (* direction of rewrite, <- *)
+    left : bool
+  }
+
+(* 
+ * Proof of x = x where x : A. 
+ *)
+type eq_refl_args = {
+    a : types;
+    x : constr;
+  }
+             
+val dest_rewrite : constr -> rewrite_args option  
```

Comment:
In addition to the `dest` functions, I usually add an `apply` function that goes back to terms